File: app/src/main/java/io/scribeapp/app/ScribeApplication.java
package io.scribeapp.app;

import android.app.Application;
import android.content.Context;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import dagger.ObjectGraph;
import io.scribeapp.connection.Session;
import io.scribeapp.input.MainInputService;
import io.scribeapp.utils.SessionLoader;

/**
 * Created by piotrekd on 12/28/13.
 */
public class ScribeApplication extends Application implements Navigator {
    private ObjectGraph objectGraph;
    private Session session;

    @Inject
    SessionLoader sessionLoader;
    private MainInputService inputServiceInstance;

    public static ScribeApplication from(Context context) {
        return (ScribeApplication) context.getApplicationContext();
    }

    @Override
    public void onCreate() {
        super.onCreate();
        objectGraph = ObjectGraph.create(getModules().toArray());
        inject(this);
    }

    public ObjectGraph getObjectGraph() {
        if (objectGraph == null) {
            objectGraph = ObjectGraph.create(getModules().toArray());
        }
        return objectGraph;
    }

    @Override
    public Session getSession() {
        if (session == null) {
            session = sessionLoader.load();
        }
        return session;
    }

    @Override
    public void setSession(Session session) {
        sessionLoader.save(session);
        this.session = session;
    }

    public Application inject(Object object) {
        objectGraph.inject(object);
        return this;
    }

    protected List<Object> getModules() {
        List<Object> modules = new ArrayList<>();
        modules.add(new ApplicationModule(this));
        return modules;
    }

    public MainInputService getInputMethodService() {
        if (inputServiceInstance == null) {
            throw new IllegalStateException("No registered input method.");
        }
        return inputServiceInstance;
    }

    public void registerInputMethodService(MainInputService ims) {
        this.inputServiceInstance = ims;
    }

    public void unregisterInputMethodService() {
        inputServiceInstance = null;
    }

    public static ScribeApplication get(Context context) {
        return (ScribeApplication) context.getApplicationContext();
    }
}

File: app/src/main/kotlin/io/scribe/app/ScribeApplication.kt
package io.scribe.app

import android.app.Application
import android.content.Context
import io.scribe.connection.Session
import io.scribe.input.MainInputService
import io.scribe.utils.SessionLoader
import io.scribe.utils.inject.RegistrableInputMethodService
import rx.Observable
import java.util.ArrayList
import javax.inject.Inject

public class ScribeApplication : BaseApplication() {
    private var session: Session? = null

    private var inputServiceInstance: RegistrableInputMethodService? = null
    var sessionLoader: SessionLoader? = null
        [Inject] set

    override fun onCreate() {
        super<BaseApplication>.onCreate()
        component.inject(this)
    }

    companion object {

        fun component(context: Context): ApplicationComponent {
            return (context.getApplicationContext() as ScribeApplication).component
        }
    }
    override fun getSession(): Session {
        if (session == null) {
            session = sessionLoader?.load()
        }
        return session!!
    }

    override fun setSession(session: Session) {
        sessionLoader?.save(session)
        this.session = session
    }

    public fun getInputMethodService(): RegistrableInputMethodService {
        if (inputServiceInstance == null) {
            throw IllegalStateException("No registered input method.")
        }
        return inputServiceInstance!!
    }

    public fun registerInputMethodService(ims: RegistrableInputMethodService) {
        this.inputServiceInstance = ims
    }

    public fun unregisterInputMethodService() {
        inputServiceInstance = null
    }
}